docs/design/proposals/stateless-csi-plugin.md

```diff
@@ -0,0 +1,511 @@
+- **volume_name**: Name of the volume as recognized by the CO system, passed in
+  only during VolumeCreate operation
+- **cluster-id**: Ceph cluster fsid
+- **pool-name**: Ceph pool on which image is created
+
+#### Name interpolation rules
+
+1. *ImageName* is known only to Ceph, *volume_id* is known only to the CO system
+1. We should be able to generate *ImageName* given a *volume_id*
+  - This is needed in most NS operations, that receive only the *volume_id*
+    and conditionally the *volume_context* and *publish_context*
+  - *volume_context* and *publish_context* are only cached, and not
+    essentially persisted, hence we cannot rely on payload from the same
+    to complete the transformation
+  - This transformation is needed by the CS specifically for DeleteVolume,
+    DeleteSnapshot and ControllerExpandVolume operations
+1. We should generate the same *ImageName*, when a Create request comes in with
+  the same *volume_name*
+  - CreateVolume may retry with the same *volume_name* on timeouts of the
+    request, and hence we need to generate the same *ImageName* for a given
+    *volume_name*
+1. We should be able to distinguish *ImageName*s that come from an instance of
+  the CSI plugin
+  - When dealing with a common Ceph cluster shared across CO systems, we need
+    the *ImageName* to be distinguishable, to avoid collisions and also to
+    be able support/aid operations like ListVolumes/Snapshots
+  - The same also applies when dealing with different CSI plugin instance in
+    the same CO system, using the same Ceph cluster
+
+#### Using Ceph OMaps for idempotence
+
+To preserve the idempotence requirement, across *ImageName* for a given
+*volume_name*, it is proposed to use a per-CSI instance Ceph omap named
+`csi_volumes_[csi-id]`, to store keys named `csi.volume.[volume_name]` with the
+CO generated *volume_name*.
+
+The value stored in `csi.volume.[volume_name]` is the *ImageName* that is to be
+used for the requested *volume_name*.
+
+A create volume request would hence operate as follows,
+
+- Use CO passed *volume_name* to generate the required key name
+- Check if key exists and read the value of the same, if it does
+- If key exists and value points to an image that satisfies the create request
+  return with success
+- If not, generate an *ImageName* and store the key and value into the omap
+- Proceed with image creation operations as required
+
+Racy key creations/accesses are prevented by locking on the *volume_name* that
+the create request carries, thus ensuring that only one such transaction for
+the key named with the *volume_name* is in flight at any given time.
+
+One other advantage of using this omap is, list operations can be performed
+more easily using the omap, rather than listing all images in the pool.
+
+#### Naming scheme
+
+ImageName format: `csi-vol-[uuid]`
+
+volume_id format: `[version]-[cluster-id]-[pool-id]-[ImageName]`
```

Comment:
Probably a stupid question, but if the ```cluster-id``` is used to look-up the mon addresses, IDs, and keys, would it not be possible just have each RBD ```StorageClass``` use its own "Secret" to encode the pool id as well (i.e. the ```cluster-id``` becomes the ```storage-class-id```)?

Reply:
... probably would help it be more consistent w/ CephFS as well since instead of a ```cluster-id``` and ```fscid``` to identify the CephFS "volume" (instead of ```pool-id```), it too could reference a unique persistent config structure.

Reply:
> Probably a stupid question, but if the `cluster-id` is used to look-up the mon addresses, IDs, and keys, would it not be possible just have each RBD `StorageClass` use its own "Secret" to encode the pool id as well (i.e. the `cluster-id` becomes the `storage-class-id`)?

The intention is to keep a cluster level configuration entity, that can be reused across pools for the same cluster. So, a single cluster ID to denote cluster level configuration, and the StorageClass carries the pool name.

If it is required that per pool user/adminID and keys be used, then these are can be skipped in the cluster configuration secret, but passed as separate secret and passed in via the csi.storage.k8s.io/*-secret-[name|namespace] options in the StorageClass.

Reply:
... and you cannot just pull the ```StorageClass``` config in the same way it is pulling the ```Secret``` (or pull a second ```Secret``` with ```StorageClass```-unique fields)? Just strikes me as odd that you need to encode the pool ID into the volume/snapshot ID but still pull the remainder of connection details from a config file.

Reply:
[Secrets](https://github.com/container-storage-interface/spec/blob/c3707a60a2d4a0a684ba342c048534defb74853d/lib/go/csi/csi.pb.go#L821) are sent in separately via the RPC call, and the [options](https://github.com/container-storage-interface/spec/blob/c3707a60a2d4a0a684ba342c048534defb74853d/lib/go/csi/csi.pb.go#L817) are sent in differently.

Post volume creation, certain calls only send in the [secret](https://github.com/container-storage-interface/spec/blob/c3707a60a2d4a0a684ba342c048534defb74853d/lib/go/csi/csi.pb.go#L1895) and not the options, and hence the pool has to be encoded in the VolumeID that the plugin uses to identify the volume.

Does this explain the observation?

Reply:
The secrets are sent or they are queried from k8s_store?

Reply:
The secrets in the keys named `csi.storage.k8s.io/*-secret-[name|namespace]` in the storage class are sent, the [secret](https://github.com/ceph/ceph-csi/blob/csi-v1.0/examples/rbd/template-ceph-cluster-ID-secret.yaml) detailing the configuration is passed to the CSI plugin, hence queried/read from k8s_store.

Reply:
OK, so it's basically a k8s built-in to send a single named secret to the controller/node agents. And because you don't want to force a 1 StorageClass-to-1 Secret mapping, the pool ID needs to get encoded in the volume/snapshot ID. However, if you did force a 1-to-1 mapping, there would be no need (similar argument for CephFS file system ID replacing pool id).

As an aside, that current "pools" parameter in the secret doesn't seem to make much sense.

Reply:
Rehashing this conversation so that I am not missing the context of the discussion:

The comment is that, we could/should move the `pool` name key out of the StorageClass and add it to the secrets referred to by a given `ClusterID`, thus removing it from being encoded in the returned `VolumeID`.

### Why the current scheme:

The current scheme is (as mentioned above) to retain a 1:1 Ceph-cluster:ClusterID configuration mapping, such that when MONs or credentials of the Ceph cluster changes, it needs to be updated in one place.

The `pools` parameter in the `clusterID` named secret, is retained for listing operations. During listing, we cannot know which pools were involved in doling out PVCs and hence which ones to list from (unless we decide to query all pools to see if the omap key is present).

### What if we move the scheme to a pool:clusterID mapping:

If we had a pool:clusterID mapping, then changes to cluster configuration needs to be edited in multiple `ClusterID` named secrets.

OTOH, listing does not need `pools` key in the `ClusterID` configuration. Also, pool need not be decoded from the VolumeID and can be referred to from the `ClusterID` configuration.

Either scheme has its pros and cons as I see it.

Thoughts?

Reply:
Pool name is retained in the storage class (and fsname for CephFS plugin will be added), as per various reviews and discussions across the implementation.